### ISSUE TITLE
Question on FixedNoiseGP.

### ISSUE THREAD

**Issue by konstantinos-p** [pytorch/botorch#769]
## Issue description

I apply the FixedNoiseGP on a training set and then evaluate on a testing set. I observe some results that are weird (to me). 1)When calculating the log likelihood of the testing data, subsets of the testing data have *lower* log likelihood that the complete testing set. For example setting

test_start=0.6
test_end=1.0
train_loc = 0.6

results in mll_test = 77. Setting

test_start=0.6
test_end=0.8
train_loc = 0.6
 
results in mll_test = 50. In general taking the average of the log likelihood of the subsets won't result in the log likelihood of the complete set. 

2)As I increase the testing set (relative to the training set) it's log likelihood also increases, with the maximum log likelihood being when the training set is 2 samples and the testing set is much larger.

I am new to Gaussian processes. Is there something that I am not understanding about the math behind the above? Am I missing some code (for example "with torch.no_grad():")? 

## Code example

```python
import torch
from botorch.models import FixedNoiseGP
from botorch.fit import fit_gpytorch_model
from gpytorch.mlls import ExactMarginalLogLikelihood
from matplotlib import pyplot as plt
from matplotlib.lines import Line2D
import numpy as np

torch.manual_seed(10)#1,5 is also good,

#Generate noisy samples with known variance
X = torch.rand(40, 1)
Y = 0.5*X
noise_weight = (X*0.4)**2
noise = torch.randn_like(Y)*torch.sqrt(noise_weight)
Y+=noise

#Divide the sample into a training and test set.
test_start=0.8
test_end=1.0
train_loc = 0.6

train_X = X[np.where(X<train_loc)[0],:]
test_X=X[np.where((X>=test_start) & (X<test_end))[0],:]
train_Y = Y[np.where(X<train_loc)[0],:]
test_Y=Y[np.where((X>=test_start) & (X<test_end))[0],:]
train_w = noise_weight[np.where(X<train_loc)[0],:]
test_w=noise_weight[np.where((X>=test_start) & (X<test_end))[0],:]

#Train the model
gp = FixedNoiseGP(train_X, train_Y, train_w)
mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_model(mll)


#Create sampling for visualization (assume that noise is perfectly known).
lin_x=torch.linspace(0, 1, 100)
lin_y = gp.likelihood(gp(lin_x),noise=(lin_x*0.4)**2)


# marginal log likelihood of training and testing data
mll_train = torch.mean(gp.likelihood(gp(torch.tensor(train_X).float()),noise=torch.tensor(train_w).float()).log_prob(torch.tensor(train_Y).float()))
mll_test = torch.mean(gp.likelihood(gp(torch.tensor(test_X).float()),noise=torch.tensor(test_w).float()).log_prob( torch.tensor(test_Y).float()))


#get the test predictive distribution for inspection
tst_distribution = gp.likelihood(gp(torch.tensor(test_X).float()),noise=torch.tensor(test_w).float())
tst_dist_mean = tst_distribution.mean.detach().numpy()
tst_dist_cov = tst_distribution.covariance_matrix.detach().numpy()

#get individual log likelihood for each testing sample for inspection
tst_log_probs = gp.likelihood(gp(torch.tensor(test_X).float()),noise=torch.tensor(test_w).float()).log_prob( torch.tensor(test_Y).float())
tst_log_probs = tst_log_probs.detach().numpy()



#get predictibe mean confidence region of predictive distribution
with torch.no_grad():
    mean = lin_y.mean.numpy()
    lower,upper = lin_y.confidence_region()

#Plot data
plt.rc('text', usetex=True)
plt.rc('font', family='serif')
plt.rcParams['text.latex.preamble']=[r"\usepackage{amsmath}"]

size_font_title = 20
size_font_legend = 15
size_font_axis = 20

fig1, ax1 = plt.subplots(figsize=(4,4))
plt.xlabel('x',fontsize=size_font_axis)
plt.ylabel('y',fontsize=size_font_axis)

ax1.set_ylim(ymin=-1, ymax=1)
#ax1.set_xlim(xmin=0, xmax=100)

colors=['#377eb8','#e41a1c','#4daf4a','#8da0cb','#ff7f00','#ffff33','#fe01b1']

area = (10 * 0.5)**2  # 0 to 15 point radii


#Plot Data Points
plt.scatter(train_X.numpy(),train_Y.numpy(),s = area,marker='o',alpha=1,color='blue')
plt.scatter(test_X.numpy(),test_Y.numpy(),s = area,marker='o',alpha=1,color='red')
plt.plot(lin_x,mean)
plt.fill_between(lin_x, lower, upper, alpha=0.5)


patch0 = Line2D([0], [0], marker='o', color='w', label='Isotropic@Zero',
       markerfacecolor=colors[0], markersize=15)
patch1 = Line2D([0], [0], marker='o', color='w', label='Isotropic@Init',
       markerfacecolor=colors[1], markersize=15)

#plt.legend(loc = 1,handles=[patch0,patch1],fontsize=size_font_legend)
plt.grid(linestyle=':')
plt.title(r'\textbf{FixedNoise}',fontsize=size_font_axis)
plt.tight_layout()



plt.matshow(tst_dist_cov[0])
```

## System Info
- botorch                   0.4.0
- gpytorch                  1.4.0
- pytorch                   1.7.1
- macos catalina

**Comment by saitcakmak** [comment]
Hi @konstantinos-p. 
While going through your code, I noticed that in the `log_prob` calls, you're passing `train_Y` which is a `6 x 1` tensor. If you take a look, you'll noice that the MVN object given by `gp(....)` or `gp.likelihood(gp(...))` calls have `MVN.loc` (mean vector) of shape `6`. This is because GpyTorch doesn't use an explicit output dimension for single output models. 
So, what ends up happening is in `log_prob` call, you have an `6 x 1`-dim input tensor `value` and a `6`-dim mean tensor `mean`. You calculate `diff = value - mean`, and use the `6x6` covariance matrix to calculate the log-probability of `diff`. When value is `6 x 1`, diff is `6 x 6`, and you get a `6`-dim output from `log_prob` call where i-th entry corresponds to the log-prob of the vector obtained by repeating `test_Y[i]` 6 times. 
To get the correct log-prob, you should pass `test_Y.squeeze(-1)` to the `log_prob` call. This will give you a single log-prob value for the 6-element vector `test_Y` under the corresponding multi-variate normal distribution (rather than 6 log-prob values for individual outputs). Hope this helps!

**Comment by Balandat** [comment]
Thanks @saitcakmak for pointing out this gotcha. @konstantinos-p I'm going to assume this resolves your question - if not please reopen the issue. 

For some more context: We chose to make output dimensions explicit in BoTorch to avoid ambiguities that can happen when you don't (e.g. is a `b x d`- dim tensor a set of `b` `d`-dim observations, or is it a `b`-sized batch of `d` scalar observations?). We internally do a bunch of shape checking and enforcing this consistently, but if you use the gpytorch methods you sometimes (as in this case) have to be careful about this.

**Comment by konstantinos-p** [comment]
Thanks alot for the quick reply and sorry for my own delay. Yes , everything seems to be consistent now. Thank you for the detailed explanations!
